avail-server/src/main/kotlin/com/avail/server/io/files/FileManager.kt

```diff
@@ -0,0 +1,394 @@
+	 *
+	 * This cache works in conjunction with [pathToIdMap] to maintain links
+	 * between
+	 */
+	// TODO make the softCapacity and strongCapacity configurable and not magic numbers
+	private val fileCache = LRUCache<UUID, MutableOrNull<ServerFileWrapper>>(
+		10000,
+		10,
+		{
+			var path = ""
+			pathToIdMap.forEach { (k, v) ->
+				if (v == it)
+				{
+					path = k
+					return@forEach
+				}
+			}
+			MutableOrNull(
+				if (path.isEmpty())
+				{
+					null
+				}
+				else
+				{
+					ServerFileWrapper(
+						it, path, openFile(Paths.get(path), fileOpenOptions))
+				})
+
+		},
+		{ _, value ->
+			try
+			{
+				value.value?.close()
+			}
+			catch (e: IOException)
+			{
+				// Do nothing
+			}
+		})
+
+	/**
+	 * Fully remove the file associated with the provided [fileCache] id. This
+	 * also removes it from [pathToIdMap].
+	 *
+	 * @param id
+	 *   The [UUID] that uniquely identifies the target file in the cache.
+	 */
+	fun remove (id: UUID)
+	{
+		fileCache[id].value?.let {
+			fileCache.remove(id)
+			pathToIdMap.remove(it.path)
+			idToPathMap.remove(id)
+		} ?: idToPathMap.remove(id)?.let { pathToIdMap.remove(it) }
+	}
+
+	/**
+	 * Deregister interest in the file associated with the provided [fileCache]
+	 * id. If the resulting [interest count][ServerFileWrapper.interestCount]
+	 * is 0, the file is closed and fully removed from the [fileCache].
+	 *
+	 * @param id
+	 *   The [UUID] that uniquely identifies the target file in the cache.
+	 */
+	fun deregisterInterest (id: UUID)
+	{
+		fileCache[id].value?.let {
+			if (it.interestCount.decrementAndGet() == 0)
+			{
+				remove(id)
+				it.close()
+			}
+		}
+	}
+
+	/**
+	 * Delete the file at the provided path.
+	 *
+	 * @param path
+	 *   The String path to the file to be deleted.
+	 * @param failure
+	 *   A function that accepts a [ServerErrorCode] that describes the nature
+	 *   of the failure and an optional [Throwable]. TODO refine error handling
+	 */
+	fun delete (
+		path: String,
+		success: () -> Unit,
+		failure: (ServerErrorCode, Throwable?) -> Unit)
+	{
+		pathToIdMap[path]?.let {id ->
+			fileCache[id].value?.delete(id, success, failure)
+			idToPathMap.remove(id)
+		} ?: {
+			if (!Files.deleteIfExists(Paths.get(path)))
+			{
+				failure(FILE_NOT_FOUND, null)
+			}
+			else
+			{
+				success()
+			}
+		}.invoke()
+	}
+
+	/**
+	 * A [Map] from the String [Path] location of a [file][AvailServerFile] to
+	 * the [UUID] that uniquely identifies that file in the [fileCache].
+	 *
+	 * This map will never be cleared of values as cached files that have been
+	 * removed from the `fileCache` must maintain association with the
+	 * server-assigned [UUID] that identifies the file for all interested
+	 * clients. If a client requests a file action with a given UUID and it is
+	 * not found in the `fileCache`, this map will be used to retrieve the
+	 * associated file from disk and placed back in the `fileCache`.
```

Comment:
If I recall correctly, I was going to tie this to open Sessions via a Session open count on the ServerFileWrapper. When a file is closed by a session, it would decrement the counter on the ServerFileWrapper. Once a wrapper's open count goes to 0, it would clean itself up from the FileManager.

I think there is enough in place, and having thought about it, I'm sure I can knock this out fairly quickly.

Reply:
Looks like I've already got this in place in SererFileWrapper via the counter, interestCount. In FileManager, there is a method, _deregisterInterest_, that is used to clean up files in the FileManager. I just need to do a little more wiring with Session to have it fully functional.